Ensure that type validation is executed for StrictMock instances as well

macisamuele opened the pull request:
<!-- 
Thank you for your interest in this project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct of this project which can be found at https://github.com/facebookincubator/TestSlide/blob/master/CODE_OF_CONDUCT.md 

Also, please make sure you're familiar with and follow the instructions in the contributing guidelines which can  be found at https://github.com/facebookincubator/TestSlide/blob/master/CONTRIBUTING.md

-->

While debugging some issue into some FB internal tests (which I was expecting failing) I've noticed that using `mock_callable` / `mock_async_callable` on `StrictMock` does not enable type validation.
This feels like an oversight on the code and has the potential to have users (myself included) writing tests expecting the framework to be validating the types (and as such ensuring that the mocks are consistent with the typing annotations).

A super small representative example is available on https://gist.github.com/macisamuele/b6e85ddfad7f5802ebedd5a2a0810c6c

This PR does also fixes/improves few more details:
* ensure that functions wrapped for type checking are not altering function name and docstring (debuggability)
* removes redundant `FORCE` Makefile target
* updates the repository link in `setup.py`

## What:

<!-- What changes are being made? (Link the feature request/issue that is being fixed here) -->


## Why:

<!-- Why are these changes necessary? -->

These changes are needed to ensure that `mock_callable` and `mock_async_callable` does preserve type validation regardless of the target being a *real* instance or a `StrictMock` instance.

**NOTE**: In order to make the change possible and well defined it does modify the `tests.sample_module` (6be4622482e77b45ce240ffbf8804ac75fbb30ae) in order to have the methods returning `Iterable`s instead of `str`s.
This is needed because current testing (`mock_callable_testslide.py`) does mock `class_method` with `to_yield_values` which is incompatible from a typing prospective. Fixing the tests would have required more work respect fixing the module class and it is safe because the module is used only for testing purposes.

## How:

<!-- How were these changes implemented? -->

The `mock_callable` implementation was intentionally excluding `StrictMock` instances to be wrapped by type validation, this PR *just* removed this exclusion.

## Risks:
<!-- Any possible risks you've likely introduced in this PR?  -->

Users of the library which have been relying on the lack of type validation on mocked functions on `StrictMock`s would see their tests failing.
This is intentional as the lack of the type validation in such cases was a bug and not considered a feature.
For these users, my suggestion would be to fix their tests or to add `type_validation=False` into the `mock_(async_)callable` definition (the later is discouraged as it hides the issue instead of solving it)

## Checklist

<!-- 
Have you done all of these things?
To check an item, place an "x" in the box like so: "- [x] Tests"
Add "N/A" to the end of each line that's irrelevant to your changes
-->


- [x] Added tests, if you've added code that should be tested
- [ ] Updated the documentation, if you've changed APIs
- [x] Ensured the test suite passes
- [x] Made sure your code lints
- [x] Completed the Contributor License Agreement ("CLA")


<!-- feel free to add additional comments -->

fornellas commented on Makefile:
```diff
@@ -36,9 +36,6 @@ endif
 .PHONY: all
 all: tests coverage_report docs sdist
 
-# .PHONY does not work for implicit rules, so we FORCE them
-FORCE:
-
```
Things are indeed worknig now, thanks for cleaning this up!

fornellas commented on testslide/lib.py:
```diff
@@ -271,6 +272,8 @@ def _wrap_signature_and_type_validation(
     skip_first_arg = _skip_first_arg(template, attr_name)
 
+    # Add this so docstrings and method name are not altered by the mock
+    @wraps(callable_template)
```
Hum... I do understand the idea here, though I'm weighting if letting it explicit "this is a mock", with new name and docstring would be better than "implicit" and keeping the old names... Not strongly opinionated either way though.

macisamuele replied:
The idea here is that I would like to prevent as much as possible to have
```
<function _wrap_signature_and_type_validation.<locals>.with_sig_and_type_validation at 0x10ee343b0>
```
as information of the wrapped function.

Doing so makes debugging very very hard (at least I had a particularly hard time to identify the mocked function in `mock_callable_testslide.py` due to the function name).

With this change the function would look like
```
<function Client.sync_function at 0x102386ef0>
```

I can see your point of preserving the information of "this is a mock", in this case we might want to have something like
```
<function TestSldeValidation(_wrap_signature_and_type_validation) at 0x105c17560>
```

If you might like this idea I can update the PR

fornellas commented:
LGTM \o/ Thanks for fixing this totally overseen case!

macisamuele commented:
I've rebased the branch on top of the latest master and addressed the suggestion about preserving "this is a mock" concept and removed a miscellaneous `facebookincubator` links

Other than that the new iteration changes the return types of `sample_module` to `List[str]` from `Iterable[str]` mostly because we would risk to have inconsistent types due to the fact that `"something"` is either a `str` and an `Iterable[str]` and this might cause confusion on tests (especially if you are expecting them failing and they are not).

Last but not least I've improved the handling of `type_validation` into `mock_callable` and `mock_async_callable`.
Before the changes (in 026af9d) we would default `type_validation` to `True` fully ignoring that the underlying `StrictMock` (if present) is defining it to `None` and the same was used to happen on the other direction where we would set `type_validation=False` into `mock_callable` definition but then type validation would still be performed.
**NOTE**: I faced this while fixing some tests internally to FB codebase and it felt interesting to fix-forward in the OS as well. The provided tests (in `tests/strict_mock_testslide.py`) highlight the issue.
In order to fix this I've decided to relax `type_validation` from a boolean to a ternary value where
* `None` means that type validation would be enable (except if the underlying strictmock would instruct otherwise)
* `True` will ensure that it is enabled
* `False` will ensure that it is disabled

I've not forgotten about this diff, but we uncovered a lot test cases breaking internally because of the new detection :)
